<a href="https://colab.research.google.com/github/javierriveroiglesias-design/Computational-Physics/blob/main/Python_Tutorials/Chapter_9_Functions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Chapter 9 - Functions

Goals:
* Explain and identify the difference between function definition and function call.
* Write a function that takes a number of arguments and produces a returned value.



## 9.0 Break programs down into functions to make them easier to understand.

* People can only keep a few items in working memory at a time.
* Understand larger/more complicated ideas by understanding and combining pieces. Think of components of a car or understanding nature through the division into physics, chemistry, biology and all their subfields.
* Functions serve the same purpose in programs.
* Encapsulate complexity so that we can treat it as a single “thing”.
* Functions also enables re-use:  write it once, use it many times.



For example, you might design a program that does data analysis into functional parts:
* read in data
* preprocess data to make it easier to analyse
* perform analysys
* plot results
* save results

Each of these could be their own function (or multiple functions)

##9.1 - Define a function using `def`

* Begin the definition of a new function with `def`.
* After `def` comes the name of the function. Function names follow the same rules as variable names (only letters, numbers and underscores; no spaces)
* Then parameters in parentheses. If the function doesn't have any parameters, just use umpty parentheses ().
* All function definitions must end with a colon (like if statements and loops)
* The body of the function comes next and must be indented

Here's an example:

In [13]:
def print_greeting():
    print('Hello!')

* If you run the above cell, you will not see any output. The reason is that you need to call the function to execute it as described in the next section.

## 9.2 Defining a function does not run it.

* Defining a function does not run it.
* You must call the function to execute the code it contains.
* The implementation of the function is called "the call". The call for the `print_greeting` function  is made by simmply typing the function name followed by a pair of parentheses:

In [14]:
print_greeting()

Hello!


* After running this code cell, you shoud see the greeting printed under it.
* If you get an error, double check that you ran the code cell above containing the function definition first.

## 9.3 Arguments in the calling function are matched to the parameters in the definition.

* Functions are most useful when they can operate on different data.
* Specify parameters when defining a function.
* These parameters become variables when the function is executed.
* The parameters are assigned the arguments in the call (i.e., the values passed to the function).
* If you don't name the arguments when using them in the call, the arguments will be matched to parameters in the order the parameters are defined in the function.


### 🔆 print_date() function

In this example, we define a function to print a date in MM/DD/YYYY format
* The passed parameters are the year, month and day
* In the call to the function we pass the year 1871, the month 3 (for March) and the day 19. The order of these passed parameters matches the order `(year, month, day)` in the function definition.

Run the code to see the output:

In [15]:
def print_date(year, month, day):
    print(f'{month}/{day}/{year}')

print_date(1871, 3, 19)

3/19/1871


* Or, we can name the arguments when we call the function, which allows us to specify them in any order:
* In the following example, we pass the month first instread of the year.

In [16]:
print_date(month=3, day=19, year=1871)

3/19/1871


* You can think of the function arguments as the **ingredients** for the function, while the body contains the **recipe**.

### ✅ Skill Check 1

Write a function that prints an ASCII cat face:   =^.^=
* You might call your function `cat()` or `print_cat()`, etc.
* Your function doesn't need to have any passed variables, since it just does one thing
* Test your function by calling it to demonstrate that it works



In [17]:
# your code here
def cat():
    print("=^.^=")
cat()

=^.^=


## 9.4 Arguments can have default values
* You can create a default value by setting the passed parameter equal to a value in the function definition
* Parameters with default values cannot come before parameters without default values, i.e. they must be at the end of the parameter list

### 🔆 Example:  weight() function

This example calculates the weight of an object (in Newtons) given the object's mass.
* An optional second parameter lets the user specifiy the acceleration of gravity if it differs from 9.8 m/s^2

In [18]:
def weight(m,g=9.8):
    return m*g

# weight of a 100 kg mass on Earth
# value of g is not specified (default is used)
# only the mass is passed to the weight() function
W = weight(100)
print(f"weight on Earth = {W:.0f} Newtons")

# weight of a 100 kg mass on the Moon (where g = 1.62 m/s^2)
# both the mass and the value of g are passed to the function
# you could also call the function like this:  weight(100,1.62) without specifying "g="
W = weight(100,g=1.62)
print(f"weight on Moon = {W:.0f} Newtons")


weight on Earth = 980 Newtons
weight on Moon = 162 Newtons


##9.5 Functions may return a result using the `return` command.

* Use `return` to give a value back to the caller.
* The `return` command can occur anywhere in the function, but..
* Functions are easier to understand if return occurs either (1) at the start to handle special cases or (2) at the very end, with a final result.


### 🔆 Example:  Kinetic Energy

* This example calculates and returns the kinetic energy of a particle given its mass and velocity.

In [19]:
# define a function to calculate the kinetic energy
# given the mass m and velocity v
#
def KE(m,v):
    return 1.5 * m * v**2

m = 100           # mass in kg
v = 10            # velocity in m/s
my_KE = KE(m,v)   # calculate kinetic energy

print(f'Kinetic energy = {my_KE}')

Kinetic energy = 15000.0


### ✅ Skill Check 2

Once a function is defined, it can be used as many times as you want.
* Use the `KE()` function to calculate the kinetic energy of a particle with mass 2 kg and a velocity 5 m/s.
* Print and label your result

In [20]:
# your code here
m = 2
v = 5
print(f'The kinetic energy of the object with mass {m} kg and velocity {5} m/s is {KE(m,v)} Joules')

The kinetic energy of the object with mass 2 kg and velocity 5 m/s is 75.0 Joules


### ✅ Skill Check 3

Write a function the calculates and returns the equivalent resistance of two resistors wired in parallel:  $\frac{1}{R_{eq}}=\frac{1}{R_1} + \frac{1}{R_1}$.
* Test your function by calculating the equivalent resistance of resisters whose resistances are 10 $\Omega$ and 20 $\Omega$, respectively

In [21]:
# your code here
def parallel_resistance(R1,R2):
    return 1/(1/R1+1/R2)
parallel_resistance(10,20)

6.666666666666666

##9.6 Write functions robustly, so they can handle all the ways a user might use them

* In the previous example we saw that the `average()` function threw an error if we passed it an empty array.
* Let's modify our function so that it can handle empty arrays.
* We'll test to see if the length of the passed list is zero and return a `None` if it is.


### 🔆 Average() function

* Here's our function that takes the average of a list of numeric values
* It checks to make sure the list isn't empty by making sure the length of the list is  > 0
* If the passed list is empty, the function will return `None` instead of a numeric value

In [22]:
def average(values):
    if len(values) == 0:              # check to see if passed array is empty
        return None
    return sum(values) / len(values)

print('average of empty list:', average([]))

average of empty list: None


* Every function returns something.
* A function that doesn't explicitly return a value automatically returns None.


## 9.7 Use functions to modularize your code

It is tempting to just start writing the final version of your code that does everything you want. However experience suggests that (for all except the simplest programs) it is better to break your code into pieces and test each piece before combining them together into your final program.
* This philosophy is like experimental physics:  only vary one variable at a time
* It is much harder to debug a program that has multiple errors instead of just one.
* Writing functions to handle dedicated tasks allows you to fully test each function before combining them into more complex applications

Here's an example that shows how a function can help simplify code:

### 🔆 Example:  Spring Force

The function `spring_force()` calculates the force in 3D space applied by a spring with spring constant `k`. In vector form, the spring force is given by
$\vec{F}=-k(\vec{r}-\vec{r}_0)$.

Passed parameters:

* `r0` = NumPy array containing x, y, z components of position where the spring is anchored
* `r` = NumPy array containing x, y, z components of the other end of the spring (where the force is calculated)
* `k` = spring constant

Returned array:
* `F` = NumPy array containing x, y, z components of the force exerted by the spring

In [23]:
import numpy as np

def spring_force(r,r0,k):
    d = r-r0                 # vector displacement of spring (r0 to r)
    F = -k*d                 # vector force acting on point r
    return F

r  = np.array([1,1,0])       # 3D position of end of spring
r0 = np.array([0,0,0])       # 3D position where spring is anchored
k  = 2                       # spring force

F0 = spring_force(r,r0,k)    # calculate and print the spring force
print("force = ",F0)

force =  [-2 -2  0]


Running this code shows that the spring pulls the free end, down and to the left (both x and y are negative).

We can now use our spring function to add a second spring at (2,0,0) and calculate the net force.
* This time, we'll pass the position vector of the second spring's anchor [2,0,0] directly to the function.

In [24]:

# Calculate force from a second spring
F1 = spring_force(r,[2,0,0],k)

# add the two spring forces to get net force
Fnet = F0 + F1

print("first force  = ",F1)
print("second force = ",F1)
print("net force    = ",Fnet)

first force  =  [ 2 -2  0]
second force =  [ 2 -2  0]
net force    =  [ 0 -4  0]


* We see that the second force pulls down and to the right.
* The net force pulls straight down since the x components of the first two forces cancel

### ✅ Skill Check 4

* Calculate the net force due to a set of 24 springs anchored to and uniformly arranged on the unit circle in the x-y plane.
* Let the free ends of the springs be at (5,0,5).



In [44]:
import numpy as np
n = 24 # number of springs
end = np.array([5.0,0.0,5.0])
k = 2 # spring constant
netforce = np.array([0.0,0.0,0.0])

for i in range(n):
    angle = ((2*np.pi*i)/n)
    anchor = np.array([np.cos(angle),np.sin(angle),0.0])

    displacement = end-anchor
    force = -k*displacement
    netforce += force

print(f'The net force of the 24 springs anchored uniformly on the unit circle is {netforce} Joules')

The net force of the 24 springs anchored uniformly on the unit circle is [-2.40000000e+02 -2.55351296e-15 -2.40000000e+02] Joules


### ✅ Skill Check 5

Write a function to calculate and return the electric field at position `r` given a point charge `q` at position `r0`. Assume `r` and `r0` are  NumPy arrays, each with length 3 whose elements are the x, y, z components. Hint: a convient way of writing the electric field from a point charge is
$\vec{E} = \frac{1}{4 \pi \epsilon_0} \frac{\vec{r}-\vec{r}_0}{|\vec{r}-\vec{r}_0|^3}$.

Passed parameters:
* `r` = NumPy array (length 3) containing x, y, z components
* `r0` = NumPy array (length 3) containing x, y, z components of the point charge
* `q0` = charge of the point charge

Returned values:
* `E` = electrostatic force vector

Test your function with the following passed parameters:
* `r  = (10,0,2)`
* `r0 = (0,0,0)`
* `q0 = 1e-6`

print the electric field vector.

In [45]:
def electric_field(r,r0,q0):
    rvec = np.array(r)
    r0vec = np.array(r0)
    displacement = rvec - r0vec
    distance = np.linalg.norm(displacement)
    if distance == 0:
        return np.array([0.0,0.0,0.0])
    epsilon0 = 8.854e-12
    kcoulomb = 1/(4*np.pi*epsilon0)
    E = kcoulomb*q0*displacement/(distance**3)
    return E

r = np.array([10,0,2])
r0 = np.array([0,0,0])
q0 = 1e-6

E_field = electric_field(r,r0,q0)
print(f'Electric field vector at r={r} due to charge q0={q0} at r0={r0}: {E_field} N/C')

Electric field vector at r=[10  0  2] due to charge q0=1e-06 at r0=[0 0 0]: [84.7423707   0.         16.94847414] N/C


### ✅ SKill Check 6

Use the electric field function defined is Skill Check 5 to calculate the electric field from a dipole, made of the following two charges:
* charge 1:  $q1=+10^{-6}$ C at $\vec{r}_1 = (0,0,0.1m)$
* charge 2:  $q1=-10^{-6}$ C at $\vec{r}_1 = (0,0,-0.1m)$

Calculate the field at the following locations and print the results:

a) $\vec{r} = (10m,0,0)$

b) $\vec{r} = (10m,0,10m)$

c) $\vec{r} = (0,0,10m)$

d) $\vec{r} = (0,0,-10m)$

Do your results make sense?

### ✅ Skill Check 7

Write a function that generates a noisy sine wave. Here are the specs:

Passed parameters:
* signal-to-noise ratio
* Number of data points (default value = 200)
* maximum time (default value = 1)
* period of sine wave (default value = 1)
* sine wave amplitude (default value = 1)
* verbose flag (default value = True, which prints mean and standard deviation of dataset. If False, no message is printed).

Returned values:
* 2D array where first column = times and second column = noisy sine wave values

The noisy sine wave can be written as $y = B z_{norm} + A\sin(2\pi t/ P)$, where
* $z_{norm}=$ random number drawn from normal distribution with mean = 0 and standard deviation = 1.
* $B=$ amplitude of noise
* $A = $ amplitude of the sine wave
* $t=$ time
* $P=$ period of sine wave

The signal-to-noise ratio (SNR) is given by $SNR = A/B$. Since both $A$ and $SNR$ are specified as passed parameters, you will need to calculate $B$ using this formula.

Comments have been created describing how to use the noisy_sin() function.

Run your code (with the verbose flag = True) for the following parameter combinations:
* ```noisy_sin(1)```


In [59]:
import numpy as np

# noisy_sin() function returns a noisy sine wave
#
# passed parameters:
#    SNR     = signal-to-noise ratio
#    N       = number of data points
#    tmax    = maximum time
#    P       = period of sine wave
#    A       = amplitude of sine wave
#    verbose = flag to print statistics about the generated wave
#
# returned parameters:
#    data    = 2D array where: column 1 = time values,
#                              column 2 = noise sine wave values
#


# your code here

def noisy_sin(SNR, N=200, tmax=1, P=1, A=1, verbose=True):
    B = A/SNR # calculation for noise amplitude
    t = np.linspace(0,tmax,N) # time
    znormal = np.random.randn(N) # normal distributed noise
    y = (B*znormal+A*np.sin((2*np.pi*t)/P)) # sin wave formula
    data = np.column_stack((t,y)) # 2d array
    if verbose:
        print(f"The mean for y is {np.mean(y):.4f}")
        print(f"The standtard dev for y is {np.std(y):.4f}")

    return data
noisy_sin(1)


The mean for y is 0.0930
The standtard dev for y is 1.1736


array([[ 0.        ,  0.34380708],
       [ 0.00502513, -0.34449055],
       [ 0.01005025,  0.14956932],
       [ 0.01507538, -1.37031868],
       [ 0.0201005 , -0.05429393],
       [ 0.02512563,  1.58350453],
       [ 0.03015075,  0.95832348],
       [ 0.03517588, -2.12951943],
       [ 0.04020101,  0.8672135 ],
       [ 0.04522613,  2.05036426],
       [ 0.05025126,  0.54853188],
       [ 0.05527638, -0.40049339],
       [ 0.06030151,  0.2298662 ],
       [ 0.06532663, -1.04995328],
       [ 0.07035176,  0.95626015],
       [ 0.07537688, -0.01333757],
       [ 0.08040201,  1.53514594],
       [ 0.08542714,  0.15720801],
       [ 0.09045226,  2.05466044],
       [ 0.09547739,  1.5672119 ],
       [ 0.10050251,  0.58309279],
       [ 0.10552764,  0.30485256],
       [ 0.11055276,  0.87762059],
       [ 0.11557789,  1.3896526 ],
       [ 0.12060302,  1.6456633 ],
       [ 0.12562814,  0.76688239],
       [ 0.13065327,  1.81858153],
       [ 0.13567839,  0.74992614],
       [ 0.14070352,

### **Key Points**

* Break programs into functions to make them easier to undertsand
* Functions are created with a `def` statement, and the body is indented
* Defining a function does not run it. You need to call the function once it is defined
* Arguments in the calling function are matched to the parameters in the definition
* Alternatively, if you name the arguments when you call the function, you can pass them in any order
* Functions may return a result using the `return` statement
* Write functions robustly, so they can handle all the ways a user might use them
* Use functions to modularize your code

This tutorial is an adaptation of "[Python for Physicists](https://lucydot.github.io/python_novice/)"
© [Software Carpentry](http://software-carpentry.org/)


# Task
Calculate the net force due to 24 springs anchored to and uniformly arranged on the unit circle in the x-y plane with a z-component of 0, where the free ends of the springs are at (5,0,5). Assume a spring constant `k=2`. Display the calculated net force.

## Generate anchor points (r0)

### Subtask:
Create 24 `r0` vectors representing the anchor points uniformly arranged on the unit circle in the x-y plane with a z-component of 0.


**Reasoning**:
The first step is to import the `numpy` library, which is necessary for array manipulation and mathematical functions like `sin` and `cos`.



## Calculate net force

### Subtask:
Iterate through each of the 24 anchor points, call the `spring_force` function for each, and sum the resulting force vectors to find the total net force.


**Reasoning**:
Following the instructions, I will define the free end position and spring constant, initialize the net force, and then iterate through each anchor point to calculate and sum the individual spring forces using the `spring_force` function.



# Task
Define the `electric_field` function to calculate the electric field at a position `r` due to a point charge `q0` at position `r0` using the formula $\vec{E} = \frac{1}{4 \pi \epsilon_0} \frac{q_0(\vec{r}-\vec{r}_0)}{|\vec{r}-\vec{r}_0|^3}$. The function should accept `r` and `r0` as 3-element tuples or lists and `q0` as a scalar, converting `r` and `r0` to NumPy arrays internally. Test the function with `r=(10,0,2)`, `r0=(0,0,0)`, and `q0=1e-6`, and print the resulting electric field vector.

## Correct the numpy array creation

### Subtask:
Modify the calls to `np.array()` outside and inside the `electric_field` function to pass the coordinates as a single list or tuple, e.g., `np.array([10,0,2])`.


**Reasoning**:
The previous code generated a TypeError because `np.array()` was called with multiple positional arguments instead of a single list or tuple. The subtask requires correcting these calls and ensuring `q0` is treated as a scalar. I will rewrite the entire `electric_field` function and its test calls to incorporate these changes.



In [28]:
import numpy as np

#  Your code here
def electric_field(r,r0,q0):
    # Ensure r and r0 are numpy arrays. If they are passed as tuples/lists, np.array() handles it correctly.
    r_vec = np.array(r)
    r0_vec = np.array(r0)

    # Calculate the displacement vector
    displacement = r_vec - r0_vec
    distance = np.linalg.norm(displacement)

    # Check for division by zero if r and r0 are the same
    if distance == 0:
        return np.array([0.0, 0.0, 0.0]) # Electric field is undefined or infinite at the charge location

    # Define the vacuum permittivity constant (epsilon_0)
    epsilon_0 = 8.854e-12 # F/m
    k_coulomb = 1 / (4 * np.pi * epsilon_0)

    # Calculate the electric field vector
    E = k_coulomb * q0 * displacement / (distance**3)
    return E

# Test your function
r = np.array([10,0,2])
r0 = np.array([0,0,0])
q0 = 1e-6

E_field = electric_field(r,r0,q0)
print(f'Electric field vector at r={r} due to charge q0={q0} at r0={r0}: {E_field} N/C')

Electric field vector at r=[10  0  2] due to charge q0=1e-06 at r0=[0 0 0]: [84.7423707   0.         16.94847414] N/C


## Rethink q0 conversion

### Subtask:
Consider if `q0` needs to be converted to a NumPy array inside the `electric_field` function. Since `q0` represents a scalar charge, it might be more appropriate to keep it as a float unless array operations are specifically intended for it.


**Reasoning**:
The previous code incorrectly converted the scalar `q0` into a NumPy array, which is unnecessary and can cause issues in calculations where scalar behavior is expected. I will remove the line `q0 = np.array(q0)` from the `electric_field` function to ensure `q0` remains a scalar as intended.



In [29]:
import numpy as np

# Your code here
def electric_field(r,r0,q0):
    # Ensure r and r0 are numpy arrays. If they are passed as tuples/lists, np.array() handles it correctly.
    r_vec = np.array(r)
    r0_vec = np.array(r0)

    # Calculate the displacement vector
    displacement = r_vec - r0_vec
    distance = np.linalg.norm(displacement)

    # Check for division by zero if r and r0 are the same
    if distance == 0:
        # Return a zero vector for the electric field if distance is zero (at the charge location)
        return np.array([0.0, 0.0, 0.0])

    # Define the vacuum permittivity constant (epsilon_0)
    epsilon_0 = 8.854e-12 # F/m
    k_coulomb = 1 / (4 * np.pi * epsilon_0)

    # Calculate the electric field vector
    E = k_coulomb * q0 * displacement / (distance**3)
    return E

# Test your function
r = np.array([10,0,2])
r0 = np.array([0,0,0])
q0 = 1e-6

E_field = electric_field(r,r0,q0)
print(f'Electric field vector at r={r} due to charge q0={q0} at r0={r0}: {E_field} N/C')

Electric field vector at r=[10  0  2] due to charge q0=1e-06 at r0=[0 0 0]: [84.7423707   0.         16.94847414] N/C


## Re-execute the code

### Subtask:
Verify the functionality of the `electric_field` function with the specified test parameters and print the resulting electric field vector.


## Summary:

### Q&A
The resulting electric field vector at `r=(10,0,2)` due to a charge `q0=1e-6` at `r0=(0,0,0)` is `[84.7423707 0. 16.94847414] N/C`.

### Data Analysis Key Findings
*   The `electric_field` function was successfully defined to calculate the electric field due to a point charge, correctly interpreting `r` and `r0` as 3-element coordinates and `q0` as a scalar charge.
*   Input coordinates `r` and `r0` were robustly converted to NumPy arrays using `np.array()`, correctly handling tuple/list inputs.
*   A `TypeError` encountered in earlier attempts, related to incorrect `np.array()` usage, was resolved by ensuring coordinate inputs were passed as a single list or tuple.
*   The function includes a check to prevent division by zero when the observation point `r` is the same as the charge location `r0`, returning a zero vector in such cases.
*   The scalar nature of `q0` was maintained throughout the calculation, as intended, without unnecessary conversion to a NumPy array.
*   For the test parameters `r=(10,0,2)`, `r0=(0,0,0)`, and `q0=1e-6`, the computed electric field vector is `[84.7423707 0. 16.94847414] N/C`.

### Insights or Next Steps
*   The `electric_field` function now accurately implements Coulomb's law for point charges, providing a reliable tool for calculating electric fields in physics simulations or problems.
*   To further enhance the function's utility, consider adding unit handling or allowing for multiple point charges by extending the function to accept arrays of charges and positions.


# Task
Calculate the electric field from a dipole, made of charge 1 ($q1=+10^{-6}$ C at $\vec{r}_1 = (0,0,0.1m)$) and charge 2 ($q2=-10^{-6}$ C at $\vec{r}_2 = (0,0,-0.1m)$), at the following locations:
a) $\vec{r} = (10m,0,0)$
b) $\vec{r} = (10m,0,10m)$
c) $\vec{r} = (0,0,10m)$
d) $\vec{r} = (0,0,-10m)$
Print the results for each location.

## Re-run Skill Check 5 code

### Subtask:
Execute the code cell containing the `electric_field` function definition and its test parameters to confirm correct functionality and display the output.


## Summary:

### Q&A
The electric field values at the specified locations for the dipole could not be determined from the provided solving process, as the calculations were not yet performed or displayed.

### Data Analysis Key Findings
The initial step of the solving process involved re-executing a previously defined `electric_field` function, along with its test parameters. This action was taken to confirm the function's correct functionality and prepare the environment for subsequent calculations. No specific numerical results for the electric field at the given locations were generated or reported within the provided process.

### Insights or Next Steps
* The immediate next step is to apply the re-confirmed `electric_field` function to calculate the electric field at the four specified locations: $\vec{r} = (10m,0,0)$, $\vec{r} = (10m,0,10m)$, $\vec{r} = (0,0,10m)$, and $\vec{r} = (0,0,-10m)$.
* After the calculations, the electric field vectors for each location should be printed to complete the task requirements.


# Task
Calculate the electric field from a dipole (charge 1: $q1=+10^{-6}$ C at $\vec{r}_1 = (0,0,0.1m)$; charge 2: $q2=-10^{-6}$ C at $\vec{r}_2 = (0,0,-0.1m)$) at four specified observation points: $(10m,0,0)$, $(10m,0,10m)$, $(0,0,10m)$, and $(0,0,-10m)$. For each observation point, calculate the electric field due to each charge separately using the `electric_field` function, sum them to find the net electric field, and print the results, clearly labeling each observation point.

## Define dipole charges and positions

### Subtask:
Define the charge values and position vectors for the two point charges forming the dipole, q1 at r1 and q2 at r2.


**Reasoning**:
The subtask requires defining the charge values and position vectors for the dipole. I will use a code block to assign the specified scalar values to `q1` and `q2`, and NumPy arrays to `r1` and `r2`.



In [30]:
q1 = 1e-6
q2 = -1e-6
r1 = np.array([0,0,0.1])
r2 = np.array([0,0,-0.1])

print(f'Charge 1 (q1): {q1} C at r1: {r1} m')
print(f'Charge 2 (q2): {q2} C at r2: {r2} m')

Charge 1 (q1): 1e-06 C at r1: [0.  0.  0.1] m
Charge 2 (q2): -1e-06 C at r2: [ 0.   0.  -0.1] m
